File: python/flink_agents/api/yaml/aliases.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
"""Static alias tables for the YAML loader.

Two tables:
- ``EVENT_ALIASES`` maps short event names to ``EVENT_TYPE`` constants.
- ``CLAZZ_ALIASES`` maps short provider names to fully-qualified class
  paths. The bucket is keyed on the resource type *and* the
  implementation language so the same alias (``ollama``) can refer to
  different classes across sections and languages.

For Java resources, the loader resolves the alias to the Java FQN and
wraps it in a Python-side wrapper class (see ``JAVA_WRAPPER_CLAZZ``).
"""

from typing import Dict

from flink_agents.api.events.chat_event import (
    ChatRequestEvent,
    ChatResponseEvent,
)
from flink_agents.api.events.context_retrieval_event import (
    ContextRetrievalRequestEvent,
    ContextRetrievalResponseEvent,
)
from flink_agents.api.events.event import InputEvent, OutputEvent
from flink_agents.api.events.tool_event import (
    ToolRequestEvent,
    ToolResponseEvent,
)
from flink_agents.api.resource import ResourceName, ResourceType
from flink_agents.api.yaml.specs import Language

EVENT_ALIASES: Dict[str, str] = {
    "input": InputEvent.EVENT_TYPE,
    "output": OutputEvent.EVENT_TYPE,
    "chat_request": ChatRequestEvent.EVENT_TYPE,
    "chat_response": ChatResponseEvent.EVENT_TYPE,
    "tool_request": ToolRequestEvent.EVENT_TYPE,
    "tool_response": ToolResponseEvent.EVENT_TYPE,
    "context_retrieval_request": ContextRetrievalRequestEvent.EVENT_TYPE,
    "context_retrieval_response": ContextRetrievalResponseEvent.EVENT_TYPE,
}

# resource_type -> language -> alias -> fully-qualified class path
CLAZZ_ALIASES: Dict[ResourceType, Dict[str, Dict[str, str]]] = {
    ResourceType.CHAT_MODEL_CONNECTION: {
        "python": {
            "ollama": ResourceName.ChatModel.OLLAMA_CONNECTION,
            "openai": ResourceName.ChatModel.OPENAI_COMPLETIONS_CONNECTION,
            "anthropic": ResourceName.ChatModel.ANTHROPIC_CONNECTION,
            "tongyi": ResourceName.ChatModel.TONGYI_CONNECTION,
            "azure_openai": ResourceName.ChatModel.AZURE_OPENAI_CONNECTION,
        },
        "java": {
            "ollama": ResourceName.ChatModel.Java.OLLAMA_CONNECTION,
            "openai_completions": ResourceName.ChatModel.Java.OPENAI_COMPLETIONS_CONNECTION,
            "openai_responses": ResourceName.ChatModel.Java.OPENAI_RESPONSES_CONNECTION,
            "anthropic": ResourceName.ChatModel.Java.ANTHROPIC_CONNECTION,
            "gemini": ResourceName.ChatModel.Java.GEMINI_CONNECTION,
            "azure_openai": ResourceName.ChatModel.Java.AZURE_OPENAI_CONNECTION,
            "azure": ResourceName.ChatModel.Java.AZURE_CONNECTION,
            "bedrock": ResourceName.ChatModel.Java.BEDROCK_CONNECTION,
        },
    },
    ResourceType.CHAT_MODEL: {
        "python": {
            "ollama": ResourceName.ChatModel.OLLAMA_SETUP,
            "openai": ResourceName.ChatModel.OPENAI_COMPLETIONS_SETUP,
            "anthropic": ResourceName.ChatModel.ANTHROPIC_SETUP,
            "tongyi": ResourceName.ChatModel.TONGYI_SETUP,
            "azure_openai": ResourceName.ChatModel.AZURE_OPENAI_SETUP,
        },
        "java": {
            "ollama": ResourceName.ChatModel.Java.OLLAMA_SETUP,
            "openai_completions": ResourceName.ChatModel.Java.OPENAI_COMPLETIONS_SETUP,
            "openai_responses": ResourceName.ChatModel.Java.OPENAI_RESPONSES_SETUP,
            "anthropic": ResourceName.ChatModel.Java.ANTHROPIC_SETUP,
            "gemini": ResourceName.ChatModel.Java.GEMINI_SETUP,
            "azure_openai": ResourceName.ChatModel.Java.AZURE_OPENAI_SETUP,
            "azure": ResourceName.ChatModel.Java.AZURE_SETUP,
            "bedrock": ResourceName.ChatModel.Java.BEDROCK_SETUP,
        },
    },
    ResourceType.EMBEDDING_MODEL_CONNECTION: {
        "python": {
            "ollama": ResourceName.EmbeddingModel.OLLAMA_CONNECTION,
            "openai": ResourceName.EmbeddingModel.OPENAI_CONNECTION,
            "tongyi": ResourceName.EmbeddingModel.TONGYI_CONNECTION,
        },
        "java": {
            "ollama": ResourceName.EmbeddingModel.Java.OLLAMA_CONNECTION,
            "bedrock": ResourceName.EmbeddingModel.Java.BEDROCK_CONNECTION,
        },
    },
    ResourceType.EMBEDDING_MODEL: {
        "python": {
            "ollama": ResourceName.EmbeddingModel.OLLAMA_SETUP,
            "openai": ResourceName.EmbeddingModel.OPENAI_SETUP,
            "tongyi": ResourceName.EmbeddingModel.TONGYI_SETUP,
        },
        "java": {
            "ollama": ResourceName.EmbeddingModel.Java.OLLAMA_SETUP,
            "bedrock": ResourceName.EmbeddingModel.Java.BEDROCK_SETUP,
        },
    },
    ResourceType.VECTOR_STORE: {
        "python": {
            "chroma": ResourceName.VectorStore.CHROMA_VECTOR_STORE,
            "mem0": ResourceName.VectorStore.MEM0_VECTOR_STORE,
        },
        "java": {
            "elasticsearch": ResourceName.VectorStore.Java.ELASTICSEARCH_VECTOR_STORE,
            "opensearch": ResourceName.VectorStore.Java.OPENSEARCH_VECTOR_STORE,
            "s3_vectors": ResourceName.VectorStore.Java.S3_VECTORS_VECTOR_STORE,
            "milvus": ResourceName.VectorStore.Java.MILVUS_VECTOR_STORE,
        },
    },
}

# Python wrapper class for each cross-language-supported resource type.
# When the user writes ``type: java``, the loader resolves the alias in
# the java bucket to a Java FQN and constructs a ResourceDescriptor whose
# ``clazz`` is the wrapper below and whose ``java_clazz`` kwarg is the
# resolved Java FQN.
JAVA_WRAPPER_CLAZZ: Dict[ResourceType, str] = {
    ResourceType.CHAT_MODEL_CONNECTION: ResourceName.ChatModel.JAVA_WRAPPER_CONNECTION,
    ResourceType.CHAT_MODEL: ResourceName.ChatModel.JAVA_WRAPPER_SETUP,
    ResourceType.EMBEDDING_MODEL_CONNECTION: ResourceName.EmbeddingModel.JAVA_WRAPPER_CONNECTION,
    ResourceType.EMBEDDING_MODEL: ResourceName.EmbeddingModel.JAVA_WRAPPER_SETUP,
    ResourceType.VECTOR_STORE: ResourceName.VectorStore.JAVA_WRAPPER_VECTOR_STORE,
}


def resolve_event_type(name: str) -> str:
    """Replace an event alias with its fully-qualified event type string,
    or pass through if no alias matches.
    """
    return EVENT_ALIASES.get(name, name)


def resolve_clazz(
    name: str, resource_type: ResourceType, language: Language = "python"
) -> str:
    """Look up ``name`` in the alias bucket for ``(resource_type, language)``.

    Returns the fully-qualified class path on hit, or ``name`` unchanged
    on miss (so users can supply a fully-qualified class path directly).
    """
    bucket = CLAZZ_ALIASES.get(resource_type, {}).get(language, {})
    return bucket.get(name, name)

File: api/src/main/java/org/apache/flink/agents/api/yaml/Aliases.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.api.yaml;

import org.apache.flink.agents.api.InputEvent;
import org.apache.flink.agents.api.OutputEvent;
import org.apache.flink.agents.api.event.ChatRequestEvent;
import org.apache.flink.agents.api.event.ChatResponseEvent;
import org.apache.flink.agents.api.event.ContextRetrievalRequestEvent;
import org.apache.flink.agents.api.event.ContextRetrievalResponseEvent;
import org.apache.flink.agents.api.event.ToolRequestEvent;
import org.apache.flink.agents.api.event.ToolResponseEvent;
import org.apache.flink.agents.api.resource.ResourceName;
import org.apache.flink.agents.api.resource.ResourceType;

import java.util.Collections;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.Map;

/**
 * Static alias tables for the YAML loader.
 *
 * <p>Two tables:
 *
 * <ul>
 *   <li>{@link #EVENT_ALIASES} maps short event names to {@code EVENT_TYPE} constants.
 *   <li>{@link #CLAZZ_ALIASES} maps short provider names to fully-qualified class paths, keyed on
 *       resource type <em>and</em> implementation language so the same alias (e.g. {@code ollama})
 *       can refer to different classes across sections and languages.
 * </ul>
 *
 * <p>For Python resources, the loader resolves the alias to the Python FQN and wraps it in a
 * Java-side wrapper class (see {@link #PYTHON_WRAPPER_CLAZZ}).
 */
public final class Aliases {

    /** Short event alias to fully-qualified {@code EVENT_TYPE} string. */
    public static final Map<String, String> EVENT_ALIASES;

    /** ResourceType to Language to alias to fully-qualified class path. */
    public static final Map<ResourceType, Map<Language, Map<String, String>>> CLAZZ_ALIASES;

    /**
     * ResourceType to Java-side wrapper FQN that embeds a Python implementation. Used when a YAML
     * resource declares {@code type: python} so the Java host wraps the Python class through an
     * existing PythonResourceWrapper implementation.
     */
    public static final Map<ResourceType, String> PYTHON_WRAPPER_CLAZZ;

    static {
        Map<String, String> ev = new HashMap<>();
        ev.put("input", InputEvent.EVENT_TYPE);
        ev.put("output", OutputEvent.EVENT_TYPE);
        ev.put("chat_request", ChatRequestEvent.EVENT_TYPE);
        ev.put("chat_response", ChatResponseEvent.EVENT_TYPE);
        ev.put("tool_request", ToolRequestEvent.EVENT_TYPE);
        ev.put("tool_response", ToolResponseEvent.EVENT_TYPE);
        ev.put("context_retrieval_request", ContextRetrievalRequestEvent.EVENT_TYPE);
        ev.put("context_retrieval_response", ContextRetrievalResponseEvent.EVENT_TYPE);
        EVENT_ALIASES = Collections.unmodifiableMap(ev);

        Map<ResourceType, Map<Language, Map<String, String>>> ca =
                new EnumMap<>(ResourceType.class);

        // CHAT_MODEL_CONNECTION
        Map<String, String> chatConnJava = new HashMap<>();
        chatConnJava.put("ollama", ResourceName.ChatModel.OLLAMA_CONNECTION);
        chatConnJava.put(
                "openai_completions", ResourceName.ChatModel.OPENAI_COMPLETIONS_CONNECTION);
        chatConnJava.put("openai_responses", ResourceName.ChatModel.OPENAI_RESPONSES_CONNECTION);
        chatConnJava.put("anthropic", ResourceName.ChatModel.ANTHROPIC_CONNECTION);
        chatConnJava.put("gemini", ResourceName.ChatModel.GEMINI_CONNECTION);
        chatConnJava.put("azure_openai", ResourceName.ChatModel.AZURE_OPENAI_CONNECTION);
        chatConnJava.put("azure", ResourceName.ChatModel.AZURE_CONNECTION);
        chatConnJava.put("bedrock", ResourceName.ChatModel.BEDROCK_CONNECTION);
        Map<String, String> chatConnPython = new HashMap<>();
        chatConnPython.put("ollama", ResourceName.ChatModel.Python.OLLAMA_CONNECTION);
        chatConnPython.put("openai", ResourceName.ChatModel.Python.OPENAI_COMPLETIONS_CONNECTION);
        chatConnPython.put("anthropic", ResourceName.ChatModel.Python.ANTHROPIC_CONNECTION);
        chatConnPython.put("tongyi", ResourceName.ChatModel.Python.TONGYI_CONNECTION);
        chatConnPython.put("azure_openai", ResourceName.ChatModel.Python.AZURE_OPENAI_CONNECTION);
        ca.put(ResourceType.CHAT_MODEL_CONNECTION, buildLangBuckets(chatConnJava, chatConnPython));

        // CHAT_MODEL
        Map<String, String> chatJava = new HashMap<>();
        chatJava.put("ollama", ResourceName.ChatModel.OLLAMA_SETUP);
        chatJava.put("openai_completions", ResourceName.ChatModel.OPENAI_COMPLETIONS_SETUP);
        chatJava.put("openai_responses", ResourceName.ChatModel.OPENAI_RESPONSES_SETUP);
        chatJava.put("anthropic", ResourceName.ChatModel.ANTHROPIC_SETUP);
        chatJava.put("gemini", ResourceName.ChatModel.GEMINI_SETUP);
        chatJava.put("azure_openai", ResourceName.ChatModel.AZURE_OPENAI_SETUP);
        chatJava.put("azure", ResourceName.ChatModel.AZURE_SETUP);
        chatJava.put("bedrock", ResourceName.ChatModel.BEDROCK_SETUP);
        Map<String, String> chatPython = new HashMap<>();
        chatPython.put("ollama", ResourceName.ChatModel.Python.OLLAMA_SETUP);
        chatPython.put("openai", ResourceName.ChatModel.Python.OPENAI_COMPLETIONS_SETUP);
        chatPython.put("anthropic", ResourceName.ChatModel.Python.ANTHROPIC_SETUP);
        chatPython.put("tongyi", ResourceName.ChatModel.Python.TONGYI_SETUP);
        chatPython.put("azure_openai", ResourceName.ChatModel.Python.AZURE_OPENAI_SETUP);
        ca.put(ResourceType.CHAT_MODEL, buildLangBuckets(chatJava, chatPython));

        // EMBEDDING_MODEL_CONNECTION
        Map<String, String> embConnJava = new HashMap<>();
        embConnJava.put("ollama", ResourceName.EmbeddingModel.OLLAMA_CONNECTION);
        embConnJava.put("bedrock", ResourceName.EmbeddingModel.BEDROCK_CONNECTION);
        Map<String, String> embConnPython = new HashMap<>();
        embConnPython.put("ollama", ResourceName.EmbeddingModel.Python.OLLAMA_CONNECTION);
        embConnPython.put("openai", ResourceName.EmbeddingModel.Python.OPENAI_CONNECTION);
        embConnPython.put("tongyi", ResourceName.EmbeddingModel.Python.TONGYI_CONNECTION);
        ca.put(
                ResourceType.EMBEDDING_MODEL_CONNECTION,
                buildLangBuckets(embConnJava, embConnPython));

        // EMBEDDING_MODEL
        Map<String, String> embJava = new HashMap<>();
        embJava.put("ollama", ResourceName.EmbeddingModel.OLLAMA_SETUP);
        embJava.put("bedrock", ResourceName.EmbeddingModel.BEDROCK_SETUP);
        Map<String, String> embPython = new HashMap<>();
        embPython.put("ollama", ResourceName.EmbeddingModel.Python.OLLAMA_SETUP);
        embPython.put("openai", ResourceName.EmbeddingModel.Python.OPENAI_SETUP);
        embPython.put("tongyi", ResourceName.EmbeddingModel.Python.TONGYI_SETUP);
        ca.put(ResourceType.EMBEDDING_MODEL, buildLangBuckets(embJava, embPython));

        // VECTOR_STORE
        Map<String, String> vsJava = new HashMap<>();
        vsJava.put("elasticsearch", ResourceName.VectorStore.ELASTICSEARCH_VECTOR_STORE);
        vsJava.put("opensearch", ResourceName.VectorStore.OPENSEARCH_VECTOR_STORE);
        vsJava.put("s3_vectors", ResourceName.VectorStore.S3_VECTORS_VECTOR_STORE);
        vsJava.put("milvus", ResourceName.VectorStore.MILVUS_VECTOR_STORE);
        Map<String, String> vsPython = new HashMap<>();
        vsPython.put("chroma", ResourceName.VectorStore.Python.CHROMA_VECTOR_STORE);
        vsPython.put("mem0", ResourceName.VectorStore.Python.MEM0_VECTOR_STORE);
        ca.put(ResourceType.VECTOR_STORE, buildLangBuckets(vsJava, vsPython));

        CLAZZ_ALIASES = Collections.unmodifiableMap(ca);

        Map<ResourceType, String> wrap = new EnumMap<>(ResourceType.class);
        wrap.put(
                ResourceType.CHAT_MODEL_CONNECTION,
                ResourceName.ChatModel.PYTHON_WRAPPER_CONNECTION);
        wrap.put(ResourceType.CHAT_MODEL, ResourceName.ChatModel.PYTHON_WRAPPER_SETUP);
        wrap.put(
                ResourceType.EMBEDDING_MODEL_CONNECTION,
                ResourceName.EmbeddingModel.PYTHON_WRAPPER_CONNECTION);
        wrap.put(ResourceType.EMBEDDING_MODEL, ResourceName.EmbeddingModel.PYTHON_WRAPPER_SETUP);
        wrap.put(ResourceType.VECTOR_STORE, ResourceName.VectorStore.PYTHON_WRAPPER_VECTOR_STORE);
        PYTHON_WRAPPER_CLAZZ = Collections.unmodifiableMap(wrap);
    }

    private Aliases() {}

    private static Map<Language, Map<String, String>> buildLangBuckets(
            Map<String, String> javaBucket, Map<String, String> pythonBucket) {
        Map<Language, Map<String, String>> out = new EnumMap<>(Language.class);
        out.put(Language.JAVA, Collections.unmodifiableMap(new HashMap<>(javaBucket)));
        out.put(Language.PYTHON, Collections.unmodifiableMap(new HashMap<>(pythonBucket)));
        return Collections.unmodifiableMap(out);
    }

    /** Look up an event alias; return {@code name} unchanged on miss. */
    public static String resolveEventType(String name) {
        return EVENT_ALIASES.getOrDefault(name, name);
    }

    /**
     * Look up a class alias for {@code (resourceType, language)}; return {@code name} unchanged on
     * miss.
     */
    public static String resolveClazz(String name, ResourceType resourceType, Language language) {
        Map<Language, Map<String, String>> byLang = CLAZZ_ALIASES.get(resourceType);
        if (byLang == null) {
            return name;
        }
        Map<String, String> bucket = byLang.get(language);
        if (bucket == null) {
            return name;
        }
        return bucket.getOrDefault(name, name);
    }
}
